### PR TITLE
auth: fix cluster init deadlock

### PR DESCRIPTION
**Pull request by dboslee**
If the bootstrapped resources are created but a failure occurs later in the cluster init process subsequent runs of the cluster initialization will fail due to the existing bootstrap resources.

changelog: Improved the cluster initialization process's ability to recovery from errors

### CODE REVIEW

**Comment by zmb3** [comment]
Is this something we can get some test coverage for?

File: lib/services/local/resource_test.go [hunk 1]
```diff
@@ -55,6 +56,34 @@ func TestCreateResourcesProvisionToken(t *testing.T) {
...
+	require.NoError(t, err)
+
+	// Check that the initial call to CreateResources creates the given resources.
+	s, err := NewClusterConfigurationService(tt.bk)
+	require.NoError(t, err)
+	err = CreateResources(ctx, tt.bk, cap)
+	require.NoError(t, err)
+	got, err := s.GetAuthPreference(ctx)
+	require.NoError(t, err)
+	require.Equal(t, got.GetType(), cap.GetType())
```
Comment (zmb3) [inline lib/services/local/resource_test.go:75]:
```suggestion
	require.Equal(t, cap.GetType(), got.GetType())
```

Your assertions are backwards - expected value first, then actual value.

File: lib/services/local/resource_test.go [hunk 2]
```diff
@@ -55,6 +56,34 @@ func TestCreateResourcesProvisionToken(t *testing.T) {
...
+
+	// Check that already exists errors are ignored and the resource is not
+	// updated.
+	cap.SetType(constants.SAML)
+	err = CreateResources(ctx, tt.bk, cap)
+	require.NoError(t, err)
+	got, err = s.GetAuthPreference(ctx)
+	require.NoError(t, err)
+	require.NotEqual(t, got.GetType(), cap.GetType())
```
Comment (zmb3) [inline lib/services/local/resource_test.go:84]: Same as above.

**Review by zmb3 — APPROVED** [review]

File: lib/auth/init_test.go [hunk 3]
```diff
@@ -737,6 +737,40 @@ func keysIn[K comparable, V any](m map[K]V) []K {
...
+	services.TrustInternal
+}
+
+func (t *failingTrustInternal) CreateCertAuthority(ctx context.Context, ca types.CertAuthority) error {
+	return trace.Errorf("error")
+}
+
+// TestInitCertFailureRecovery ensures the auth server is able to recover from
+// a failure in the cert creation process.
+func TestInitCertFailureRecovery(t *testing.T) {
```
Comment (bl-nero) [inline lib/auth/init_test.go:750]: How is this test related to the fix? If we are using a fake `TrustInternal` implementation, aren't we faking something that _calls_ `CreateResources`? Does this test fail without the fix?

Comment (dboslee) [reply lib/auth/init_test.go:750]: The issue was multiple calls to `CreateResources` with the same set of resources would fail. This meant that if there was an error later in the init process that caused Teleport to crash and the init process the be retried it would get stuck on `CreateResources` on all subsequent attempts.

Its not a fake `TrustInternal`, I just wrap it to inject an error in cert creation to force a failure on the first init and then assert that a subsequent init is still able to succeed. And yes this test fails without this fix.

**Review by bl-nero — APPROVED** [review]
